### DISCUSSION TITLE
Custom result class name for LIST subquery

### DISCUSSION THREAD

**Discussion by frankvollebregt** [simolus3/drift#2405] [Q&A] [answered]
I have an SQL query that uses `LIST()` to get the result of a subquery as a List in Dart. This works nicely, but the resulting class is auto-generated, and I am wondering whether I can tell the `build_runner` which name to use for this class.

```sql
getPersonsWithCars: SELECT
  persons.name,
  persons.age,
  LIST(SELECT brand, model FROM cars JOIN ownership ON cars.id = ownership.car_id WHERE ownership.person_id = persons.id)
FROM persons;
```
This query would retrieve each person, and for each person get a list of all of the `car`s owned by that person (by joining via the `ownership` table.

Now if I want to name the result class of the query, or name the resulting List of the `LIST` subquery, that's quite straightforward:
```sql
getPersonsWithCars AS PersonWithCars: SELECT
  persons.name,
  persons.age,
  LIST(SELECT brand, model FROM cars JOIN ownership ON cars.id = ownership.car_id WHERE ownership.person_id = persons.id) AS cars
FROM persons;
```
But the type class for `cars` in the `PersonWithCars` class will now be automatically generated to be `GetPersonsWithCarsCars`. Instead, I'd like to define the name of this class myself. Is this possible with Drift and SQL files?

Or, if there is a better approach to retrieve such data in general, please let me know that too ;)

**Comment by simolus3** [comment, marked as answer]
No, at the moment this is not possible. It shouldn't be too hard to implement, but I'm worried about how this could be pulled of with  reasonable syntax (`LIST (...) AS x` is already taken).

I haven't tried it, but you might be able to hack this together by defining the subquery (without the `WHERE`) as a view and then changing the `LIST` into a `SELECT * FROM that_view WHERE ... = persons.id)`. That should cause the list column to take the type of the view, and that can be changed :D But it's a lot of stuff you don't need for a simple name change.

I'm open for ideas on good syntax here, which is probably the main challenge for this feature.

**Reply by frankvollebregt** [reply]
Thank you for your reply. I agree that the syntax would quickly become confusing. The only thing I can come up with right now is using a Kotlin-like type annotation, so it would become something like
```sql
LIST(...) AS list_name: ClassName
```
to annotate that the variable is named `list_name` and values should be of type `ClassName`, but already as I write this it seems a bit confusing :p

I'll just deal with the sort of confusing name for now (though I appreciate your proposed hack using a view), and if I have any nice ideas for the syntax here I'll let you know.
